subcommand/mesh-init/checks.go

```diff
@@ -42,10 +55,169 @@ func (c *Command) constructChecks(service *api.AgentService) api.HealthChecks {
+			}
+
+			if err != nil {
+				c.log.Warn("failed to update Consul health status", "err", err)
+			} else {
+				c.log.Info("container health check updated in Consul",
+					"name", container.Name,
+					"status", container.Health.Status,
+					"statusSince", container.Health.StatusSince,
+					"exitCode", container.Health.ExitCode,
+				)
+				currentStatuses[container.Name] = container.Health.Status
+			}
+		}
+	}
+
+	return currentStatuses
+}
+
+// setChecksCritical sets checks for all of the containers to critical
+func (c *Command) setChecksCritical(consulClient *api.Client, taskID, serviceName, clusterARN string, parsedContainerNames []string) error {
+	var result error
+
+	for _, containerName := range parsedContainerNames {
+
+		var err error
+		if containerName == config.ConsulDataplaneContainerName {
+			err = c.handleHealthForDataplaneContainer(consulClient, taskID, serviceName, clusterARN, containerName, ecs.HealthStatusUnhealthy)
+		} else {
+			checkID := constructCheckID(makeServiceID(serviceName, taskID), containerName)
+			err = c.updateConsulHealthStatus(consulClient, checkID, clusterARN, ecs.HealthStatusUnhealthy)
+		}
+
+		if err == nil {
+			c.log.Info("set Consul health status to critical",
+				"container", containerName)
+		} else {
+			c.log.Warn("failed to set Consul health status to critical",
+				"err", err,
+				"container", containerName)
+			result = multierror.Append(result, err)
+		}
+	}
+
+	return result
+}
+
+// handleHealthForDataplaneContainer takes care of the special handling needed for syncing
+// the health of consul-dataplane container. We register two checks (one for the service
+// and the other for proxy) when registering a typical service to the catalog. Updates
+// should also happen twice in such cases.
+func (c *Command) handleHealthForDataplaneContainer(consulClient *api.Client, taskID, serviceName, clusterARN, containerName, ecsHealthStatus string) error {
+	var checkID string
+	var err error
+	serviceID := makeServiceID(serviceName, taskID)
+	if c.config.IsGateway() {
+		checkID = constructCheckID(serviceID, containerName)
+		err = c.updateConsulHealthStatus(consulClient, checkID, clusterARN, ecsHealthStatus)
+
+		return err
+	}
+
+	checkID = constructCheckID(serviceID, containerName)
+	err = c.updateConsulHealthStatus(consulClient, checkID, clusterARN, ecsHealthStatus)
+	if err != nil {
+		return err
+	}
+
+	proxySvcID, _ := makeProxySvcIDAndName(serviceID, "")
+	checkID = constructCheckID(proxySvcID, containerName)
+	err = c.updateConsulHealthStatus(consulClient, checkID, clusterARN, ecsHealthStatus)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Command) updateConsulHealthStatus(consulClient *api.Client, checkID string, clusterARN string, ecsHealthStatus string) error {
+	consulHealthStatus := ecsHealthToConsulHealth(ecsHealthStatus)
+
+	check, ok := c.checks[checkID]
+	if !ok {
+		return fmt.Errorf("unable to find check with ID %s", checkID)
+	}
+
```

Comment:
Should this function should return if `consulHealthStatus` == `check.Status`?

Reply:
It won't reach this place if the container's health status is same as the consul check's status. https://github.com/hashicorp/consul-ecs/blob/067dd3f01f2c48ec61c7daebb584958e54c85887/subcommand/mesh-init/checks.go#L111